testing/opae-c/test_feature_enum_c.cpp

```diff
@@ -0,0 +1,257 @@
+		ASSERT_EQ(fpgaPropertiesSetObjectType(filter_, FPGA_ACCELERATOR), FPGA_OK);
+		num_matches_ = 0;
+		ASSERT_EQ(fpgaEnumerate(&filter_, 1, tokens_.data(), tokens_.size(),
+		  &num_matches_), FPGA_OK);
+		EXPECT_EQ(num_matches_, platform_.devices.size());
+		accel_ = nullptr;
+		ASSERT_EQ(fpgaOpen(tokens_[0], &accel_, 0), FPGA_OK);
+		system_->register_ioctl_handler(FPGA_PORT_GET_REGION_INFO, mmio_ioctl);
+		which_mmio_ = 0;
+		uint64_t *mmio_ptr = nullptr;
+		EXPECT_EQ(fpgaMapMMIO(accel_, which_mmio_, &mmio_ptr), FPGA_OK);
+		EXPECT_NE(mmio_ptr, nullptr);
+
+		feature_filter_.type = FPGA_DMA_FEATURE;
+		fpga_guid guid = {0xE7, 0xE3, 0xE9, 0x58, 0xF2, 0xE8, 0x73, 0x9D, 
+					0xE0, 0x4C, 0x48, 0xC1, 0x58, 0x69, 0x81, 0x87 };
+		memcpy_s(feature_filter_.guid, sizeof(fpga_guid), guid, sizeof(fpga_guid));
+	}
+
+	void DestroyTokens() {
+		for (auto &t : tokens_) {
+			if (t) {
+				EXPECT_EQ(fpgaDestroyToken(&t), FPGA_OK);
+				t = nullptr;
+			}
+		}
+		num_matches_ = 0;
+	}
+
+	virtual void TearDown() override {
+		DestroyTokens();
+		if (filter_ != nullptr) {
+			EXPECT_EQ(fpgaDestroyProperties(&filter_), FPGA_OK);
+		}
+		system_->finalize();
+	}
+
+	uint32_t which_mmio_;
+	std::array<fpga_feature_token, 2> ftokens_;
+	fpga_handle accel_;
+	fpga_feature_properties feature_filter_;
+
+	fpga_properties filter_;
+	std::array<fpga_token, 2> tokens_;
+	uint32_t num_matches_;
+	test_platform platform_;
+	test_device invalid_device_;
+	test_system *system_;
+};
+
+/**
+ * @test       test_feature_enumerate
+ * @brief      Tests: fpgaFeatureEnumerate, xfpga_fpgaCloneFeatureToken
+ * @details    When fpgaFeatureEnumerate() is called with a valid param,<br>
+ *             then it enumerates the accelerateor's DFH to find specific<br>
+ *             type of BBB specified in the feature filter<br>
+ *             When xfpga_fpgaCloneFeatureToken() is called with a valid param,<br>
+ *             fpga_feature_token is cloned.<br>
+ */
+TEST_P(opae_feature_enum_c_p, test_feature_enumerate) {
+
+	struct DFH dfh ;
+	dfh.id = 0x1;
+	dfh.revision = 0;
+	dfh.next_header_offset = 0x100;
+	dfh.eol = 1;
+	dfh.reserved = 0;
+	dfh.type = 0x1;
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x0, dfh.csr));
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x8, 0xf89e433683f9040b));
+	EXPECT_EQ(FPGA_OK,fpgaWriteMMIO64(accel_, which_mmio_, 0x10, 0xd8424dc4a4a3c413));
+
+	struct DFH dfh_bbb = { 0 };
+
+	dfh_bbb.type = 0x2;
+	dfh_bbb.id = 0x2;
+	dfh_bbb.revision = 0;
+	dfh_bbb.next_header_offset = 0x000;
+	dfh_bbb.eol = 1;
+	dfh_bbb.reserved = 0;
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x100, dfh_bbb.csr));
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x108, 0x9D73E8F258E9E3E7));
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x110, 0x87816958C1484CE0));
+
+	EXPECT_EQ(fpgaFeatureEnumerate(accel_, &feature_filter_, NULL,
+		0, &num_matches_), FPGA_OK);
+
```

Comment:
Need to verify that num_matches_ is what you expect it to be, based on the setup you've done.